### PR TITLE
Upgrade Rust to build Cargo

### PR DESCRIPTION
**Pull request by alexcrichton**
A few small tweaks to tests were necessary to accomodate rust-lang/rust#29520,
and a few changes were made to the code to account for that as well.

### CODE REVIEW

File: tests/test_cargo_compile_custom_build.rs [hunk 1]
```diff
@@ -1135,7 +1135,8 @@ test!(build_script_with_dynamic_native_dependency {
         "#)
         .file("bar/src/lib.rs", r#"
             pub fn bar() {
-                #[link(name = "builder")]
+                #[cfg_attr(not(target_env = "msvc"), link(name = "builder"))]
+                #[cfg_attr(target_env = "msvc", link(name = "builder.lib"))]
```
Comment (retep998) [inline tests/test_cargo_compile_custom_build.rs]: Shouldn't this be `"builder.dll"` since the library is `builder.dll.lib`?

Comment (alexcrichton) [inline tests/test_cargo_compile_custom_build.rs]: I believe so, yes

**Pull request merged**